### PR TITLE
Added ship_traffic AIS example

### PR DESCRIPTION
**Pull request by jbednar**
This notebook/dashboard shows USA AIS vessel (typically ship) tracking data collected from [marinecadastre](https://marinecadastre.gov/ais/), colored by vessel type:
 
Screen Shot 2021-01-18 at 7 58 33 PM
Screen Shot 2021-01-18 at 8 03 10 PM
Screen Shot 2021-01-18 at 8 04 50 PM
Screen Shot 2021-01-18 at 8 06 05 PM
Screen Shot 2021-01-18 at 8 06 53 PM

Apart from the actual data shown, which is interesting enough, the notebook also shows:
- [x] How to process a large set of CSV files efficiently, caching result to a fast and efficient Parquet file.
- [x] How to create and persist spatial indexing for SpatialPandas
- [x] How to plot categorical data using Datashader
- [x] How to select and highlight individual data points in a datashaded plot
- [x] How to drill down into a Datashaded plot to reveal metadata about each point not shown in the plot itself

To do items:
- [x] Add a series of static zoomed images like in this issue, to show data at each scale. 
- [x] Highlight some interesting bits of the data
- [x] The spatial points selection code isn't currently working; I believe @jlstevens can replace it with some that he has.
- [x] I'm not sure the spatial indexing in general is up and running; it's not fast per zoom even if only zooming slightly more in.
- [x] After selecting a point, it would be useful to update a short but wide table of data below the plot to show that row of data in the original dataset
- [x] It would also be great to be able to use show additional metadata about each selected point:
  - [x] We would first need to build a table with metadata about the vessels covered by this dataset, by reading the data and collecting the latest non-NaN value for `['imo', 'call_sign', 'vessel_name', 'vessel_type','length', 'width']` for each `'mmsi_id'`. This data shouldn't vary per ping, and can be stored separately from the ping data.
  - [x] Once a point has been selected, look up this vessel info in a separate table
- [ ] It would also be great to show connected tracks:
  - [ ] Highlight a specific track to see how things connect, on top of a datashaded points plot
  - [ ] Datashade the whole set of pings, which might be slow and also may be confusing when it connects distant points -- may need to filter out gaps.
- [ ] Maybe selection can be in one of several modes: select nearest (default; highlights one point with a circle and shows one row in a table), select nearest N (highlights several nearby points and shows a table with multiple rows, perhaps with hovering over one point highlighting that row in the table to disambiguate), select all points with same column value as the selected one (MMSI in this case; highlights all such points and shows a table with multiple rows (up to a max), same but connecting each point into a trajectory by datetime
- [x] Once info is available per point, need to collect interesting examples and talk about them in the notebook
- [x] Test deployed app - responsive like in the notebook?
- [ ] Add dataset download info to anaconda-project.yml, putting the data (e.g. vessel metadata and Parquet file of points) on S3?

### CODE REVIEW

**Comment by jlstevens** [comment]
Very nice! This data generates some very pretty plots.

> How to select and highlight individual data points in a datashaded plot
>  How to drill down into a Datashaded plot to reveal metadata about each point not shown in the plot itself

I'm currently working on different approaches to do this. Right now I have one prototype (that needs to be better encapsulated) but I am also working on a more efficient approach which will probably be needed for this large dataset.

File: ship_tracking/ship_tracking.ipynb [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    "basedir = './2020/'\n",
+    "basename = 'AIS_2020_01'\n",
+    "\n",
+    "def load_data():\n",
+    "    cache_file = basedir+basename+\"_gdf.parquet\"\n",
+    "    if os.path.exists(cache_file):\n",
+    "        print(\"Reading Parquet file\")\n",
+    "        return sp.io.read_parquet_dask(cache_file).persist()\n",
+    "\n",
+    "    df = dd.read_csv(basedir+'AIS_2020_01*.csv')\n",
```
Comment (philippjfr) [inline ship_tracking/ship_tracking.ipynb:164]:
```suggestion
    "    df = dd.read_csv(basedir+basename+'*.csv', usecols=['MMSI', 'LON', 'LAT', 'VesselType'])
\n",
```

Comment (philippjfr) [reply ship_tracking/ship_tracking.ipynb:164]: Getting another speedup with this, now looks like this on my MBP:

Screen Shot 2021-01-19 at 12 37 43 PM

File: ship_tracking/ship_tracking.ipynb [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+   "outputs": [],
+   "source": [
+    "basedir = './2020/'\n",
+    "basename = 'AIS_2020_01'\n",
+    "\n",
+    "def load_data():\n",
+    "    cache_file = basedir+basename+\"_gdf.parquet\"\n",
+    "    if os.path.exists(cache_file):\n",
+    "        print(\"Reading Parquet file\")\n",
+    "        return sp.io.read_parquet_dask(cache_file).persist()\n",
```
Comment (philippjfr) [inline ship_tracking/ship_tracking.ipynb:162]: For a reason I can't fathom this increases rendering performance by 10x for me:

```suggestion
    "        return sp.io.read_parquet_dask(cache_file).persist().persist()\n",
```

File: ship_tracking/ship_tracking.ipynb [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    "\n",
+    "    df = dd.read_csv(basedir+'AIS_2020_01*.csv')\n",
+    "    with ProgressBar():\n",
+    "        print(\"Reading CSVs\")\n",
+    "        gdf = df.map_partitions(convert_partition, meta=example).persist()\n",
+    "        print(\"Writing Parquet file\")\n",
+    "        return gdf.pack_partitions_to_parquet(cache_file, npartitions=16).persist()"
```
Comment (philippjfr) [inline ship_tracking/ship_tracking.ipynb:169]: I think we want more partitions after all, it's still 15-25 million rows per partition here.

**Comment by jlstevens** [comment]
Together with @philippjfr, we identified the reason for the massive performance difference: you want to use `pyarrow>=2` otherwise things are exceedingly slow.

**Comment by philippjfr** [comment]
>Together with @philippjfr, we identified the reason for the massive performance difference: you want to use pyarrow>=2 otherwise things are exceedingly slow (when reading parquet).

Also writing the parquet file, it goes from 30 minutes -> 10 minutes, while reading goes from 10 minutes -> 10 seconds.

File: ship_tracking/ship_tracking.ipynb (hunk 3 above)
Comment (jbednar) [inline ship_tracking/ship_tracking.ipynb:170]: Ok:
```suggestion
    "        return gdf.pack_partitions_to_parquet(cache_file).persist()"
```

**Comment by jlstevens** [comment]
I'm guessing the data is from [this page](https://coast.noaa.gov/htdata/CMSP/AISDataHandler/2020/index.html) and obtained using:

```
wget -np -r -nH -L --cut-dirs=3 https://coast.noaa.gov/htdata/CMSP/AISDataHandler/2020/
```

And 49.3GB in size.

**Comment by jlstevens** [comment]
In 136c358 I added a (slow! non spatially indexed!) prototype of the `inspect_points` operation using https://github.com/holoviz/holoviews/pull/4796 to also make use of the changes in https://github.com/holoviz/holoviews/pull/4792

**Comment by jlstevens** [comment]
@jbednar Now you can use https://github.com/holoviz/holoviews/pull/4794 which is branched off master instead of holoviz/holoviews#4796

**Pull request merged**